component/grant-type/src/main/java/org/wso2/carbon/identity/oauth2/grant/token/exchange/TokenExchangeGrantHandler.java

```diff
@@ -0,0 +1,272 @@
+                        + "or Audience) are empty in the given JWT");
+            }
+            identityProvider = TokenExchangeUtils.getIdPByIssuer(jwtIssuer, tenantDomain);
+            tokenEndPointAlias = TokenExchangeUtils.getTokenEndpointAlias(identityProvider, tenantDomain);
+            try {
+                if (signedJWT != null) {
+                    if (TokenExchangeUtils.validateSignature(signedJWT, identityProvider, tenantDomain)) {
+                        log.debug("Signature/MAC validated successfully.");
+                    } else {
+                        handleException(OAuth2ErrorCodes.INVALID_REQUEST, "Signature or Message Authentication " +
+                                "invalid");
+                    }
+                }
+            } catch (JOSEException e) {
+                handleException(OAuth2ErrorCodes.INVALID_REQUEST, "Error when verifying signature");
+            }
+            TokenExchangeUtils.setAuthorizedUser(tokReqMsgCtx, identityProvider, subject);
+
+            log.debug("Subject(sub) found in JWT: " + subject);
+            log.debug(subject + " set as the Authorized User.");
+
+            tokReqMsgCtx.setScope(tokReqMsgCtx.getOauth2AccessTokenReqDTO().getScope());
+            if (StringUtils.isEmpty(tokenEndPointAlias)) {
+                handleException(OAuth2ErrorCodes.INVALID_REQUEST, "Token Endpoint alias of the local Identity " +
+                        "Provider has not been configured for " + identityProvider.getIdentityProviderName());
+            }
+
+            audienceFound = validateAudience(audiences, tokenEndPointAlias, requested_audience);
+            if (!audienceFound) {
+                handleException(TokenExchangeConstants.INVALID_TARGET, "Invalid audience values provided");
+            }
+            TokenExchangeUtils.checkExpirationTime(expirationTime, currentTimeInMillis, timeStampSkewMillis);
+            if (notBeforeTime != null) {
+                TokenExchangeUtils.checkNotBeforeTime(notBeforeTime, currentTimeInMillis, timeStampSkewMillis);
+            } else {
+                log.debug("Not Before Time(nbf) not found in JWT. Continuing Validation");
+            }
+            if (issuedAtTime == null) {
+                log.debug("Issued At Time(iat) not found in JWT. Continuing Validation");
+            } else if (!validateIAT) {
+                log.debug("Issued At Time (iat) validation is disabled for the JWT");
+            } else {
+                TokenExchangeUtils.checkValidityOfTheToken(issuedAtTime, currentTimeInMillis, timeStampSkewMillis,
+                        validityPeriod);
+            }
+            boolean customClaimsValidated = validateCustomClaims(customClaims);
+            if (!customClaimsValidated) {
+                handleException(OAuth2ErrorCodes.INVALID_REQUEST, "Custom Claims in the JWT were invalid");
+            }
+            log.debug("Subject JWT Token was validated successfully");
```

Comment:
add a check to see if log debug is enabled.

Reply:
do we ned here, given that we aren't doing any string concatenation

Reply:
Yes, we don't need to have the debug enabled check here.